### PR TITLE
[MANA-59] Implement dynamic memory region allocation for mtcp_restart.c

### PR DESCRIPTION
**Pull request by fyshhh**
This pull request enables `mana_restart` functionalities on CentOS. Notably, it dynamically determines whether there are vvar and vDSO segments that need to be temporarily remapped, and finds regions to map them to. This also removes some of the hard-coding done when reserving memory regions, as well as slightly changing the implementation of determining the regions to reserve.

This has been tested on both Cori and CentOS by running all the programs in `contrib/mpi-proxy-split/test` directories, and the results seen are the same as when running `mana_launch`. I also checked the addresses assigned for specific cases to ensure that all the code is necessary.

Eventually, we should seek to remove all uses of the `GNI` and `MANA_USE_LH_FIXED_ADDRESS` macros, but this is a good stopgap.

### CODE REVIEW

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [inline src/mtcp/mtcp_restart.c:685]: i suspect this is now obsolete.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [inline src/mtcp/mtcp_restart.c:750]: this (and line 751, i can't include it in the highlight) should also be obsolete. `mygetauxval` and `mysetauxval` are only called here.

**Comment by fyshhh** [comment]
there are several comments left by twinkle and replies made by me in the code. they're only here while the code is under review - once it is ready for merge, they'll either be removed or succinctly summarized to explain certain design decisions.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:528]: use `PAGESIZE`, avoid using magic number as much as possible.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:525]: Why choose 5 and 3 here? Please add comments to explain that

And if they're always 2 pages apart, you only need 1 variable I think.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:689]: First, I think we can use one function to do them all, it doesn't make too much sense to me to separate them.

Second, why do we need to remap the vvar and vdso? Please add comments to explain.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:787]: I would suggest changing the design like this:
1. change the function name to `reserve_memory_regions`
2. In the implementation of this function, use `#ifdef` to have different implementations for different cases,
```
#ifdef GNI
// reserve the memroy regions
#endif
// do nothing for other cases
```
This would be much more cleaner and general, and easier to expand if we need to add a new case.
https://en.wikipedia.org/wiki/Open–closed_principle

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:795]: ditto

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:74]: Line 68, I think we shouldn't have `GNI` when building on centos.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:787]: yes, i was speaking to twinkle about this, and this function should be working on CentOS as well. i'll try to debug it with the goal of generalizing it such that we don't require an `#ifdef` clause.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:689]: 1. that's true.
2. noted.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:525]: using two might make it clearer for the reader, especially later on when we are comparing addresses. agreed on the additional comments, though.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:74]: hmm, what do you think @JainTwinkle?

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:544]: One of the reasons why I suggested to combine them in the same function is because I don't think it necessary to re-open the same file and re-iterate again.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:559]: Can we assume there will be only 1 vvar and 1 vdso? Then we probably want to break the loop earlier.

The entries in the `/proc/self/maps` could be very large if it uses a lot of memory.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:562]: do we want to abort to return an error here?

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:559]: i have a break condition, but it may not work if there isn't a vvar (like on CentOS).

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:562]: yes, i think we should, but i would defer to the opinions of @gc00 and @JainTwinkle.

File: contrib/mpi-proxy-split/Makefile
Comment (jungan) [inline contrib/mpi-proxy-split/Makefile:113]: Please rebase, don't include unrelated commits.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:538]: if it's `vvar`, it won't be possible to be `vdso`, right? why check it twice?

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:563]: I guess it should be `(vdsoStart > 0 && vdsoStartTmp == 0)` mean?

BTW, you can just check the "temp region" at line 569 and 580.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:573]: Usually, we want to check it right after we assign it.

When `vvarStart == 0`, it makes no sense to check it, and easy to break when someone else need to add something in this function.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:563]: yes, you're right. thanks for catching it.

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:745]: I'm curious when will we go this routine?
`(libsEnd + 1 * GB < highMemStart)`

File: src/mtcp/mtcp_restart.c
Comment (jungan) [inline src/mtcp/mtcp_restart.c:779]: It looks like `start2` and `end2` are always 0? Then why do we need them?

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:745]: this happens on Cori.

File: src/mtcp/mtcp_restart.c
Comment (JainTwinkle) [reply src/mtcp/mtcp_restart.c:779]: @jungan and @fyshhh, I'm not very sure either. But, I guess the intention was to show that if we want to block/reserve more than one region, then you can have another pair of `start` and `end`. This would be helpful if you want to reserve two small memory regions instead of one large region. Or if you have some restrictions in blocking one large region and you have to reserve in parts.
I believe @gc00  and Yao kept this code if at all they need it in the future but never used it.
**The question here is:** 
1. Do we need this code block or not?
2. Can we add a comment to describe the intention/purpose of `start2` and `end2`?

File: src/mtcp/mtcp_restart.c
Comment (JainTwinkle) [inline src/mtcp/mtcp_restart.c:538]: Why are we assuming the `vvar` region's size here? 
The `/proc/maps` should give us the exact size of each region on that platform?

File: src/mtcp/mtcp_restart.c
Comment (jungan) [reply src/mtcp/mtcp_restart.c:779]: Yeah, if would be helpful if we can have some comments here.

And If we really want to have some flexibility here, I think we can use a list of addresses.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:538]: hmm, that's true. i think we can try finding the regions and their respective sizes first, then reset the file pointer and look for empty regions to map them too.

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:779]: i think we can consider removing the entire block and maybe think of making it open for extension (like jun's suggestion). this is so that in the future, if we need three, or four, or more small regions, we don't just append them to the arguments. we can also use `stdarg` instead, although that might be less safe.

File: src/mtcp/mtcp_restart.c
Comment (JainTwinkle) [inline src/mtcp/mtcp_restart.c:502]: Could you elaborate that why the global variable would persist if the `unmap_memory_areas_and_restore_vdso` function is unmapping all mtcp regions including heap and data sections?

File: src/mtcp/mtcp_restart.c
Comment (JainTwinkle) [reply src/mtcp/mtcp_restart.c:779]: Yes, I agree with removing these extra arguments. And, I'd suggest either changing it to a list or having a comment that currently, we reserve only one large area. If someone needs to extend it in the future then they can use it as a list (and give an example function signature).
@gc00 any comments or suggestions?

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:502]: i'm not too sure what you mean by this. i'm using global variables here because another function uses it later on.

Comment (JainTwinkle) [reply src/mtcp/mtcp_restart.c:502]: It is a bit informal to say "I think we're forced". I meant to update the comment stating the reason instead of a dubious statement. Rest LGTM :)

**Review by JainTwinkle — APPROVED** [review]
Please address the last comment and rest LGTM :)

**Review by jungan — APPROVED** [review]

File: src/mtcp/mtcp_restart.c
Comment (fyshhh) [reply src/mtcp/mtcp_restart.c:502]: ah ok, very true.

**Pull request merged**
